source/extensions/common/wasm/v8/v8.cc

```diff
@@ -552,39 +562,49 @@ template <typename... Args>
 void V8::registerHostFunctionImpl(absl::string_view moduleName, absl::string_view functionName,
                                   void (*function)(void*, Args...)) {
   ENVOY_LOG(trace, "[wasm] registerHostFunction(\"{}.{}\")", moduleName, functionName);
+  auto data = std::make_unique<FuncData>(absl::StrCat(moduleName, ".", functionName));
   auto type = wasm::FuncType::make(convertArgsTupleToValTypes<std::tuple<Args...>>(),
                                    convertArgsTupleToValTypes<std::tuple<>>());
   auto func = wasm::Func::make(
       store_.get(), type.get(),
       [](void* data, const wasm::Val params[], wasm::Val[]) -> wasm::own<wasm::Trap*> {
+        auto func_data = reinterpret_cast<FuncData*>(data);
+        ENVOY_LOG(trace, "[wasm] callHostFunction(\"{}\")", func_data->name);
         auto args_tuple = convertValTypesToArgsTuple<std::tuple<Args...>>(params);
         auto args = std::tuple_cat(std::make_tuple(current_context_), args_tuple);
-        auto function = reinterpret_cast<void (*)(void*, Args...)>(data);
+        auto function = reinterpret_cast<void (*)(void*, Args...)>(func_data->raw_func);
         absl::apply(function, args);
         return nullptr;
       },
-      reinterpret_cast<void*>(function));
-  host_functions_.emplace(absl::StrCat(moduleName, ".", functionName), std::move(func));
+      data.get());
```

Comment:
You don't want to use a more sophisticated constructor and emplace?  Trying to avoid piecewise constructors?

Reply:
You could also use default constructor and a copy constructor.

Reply:
Which object are you talking about? `FuncData`? It's kind of ugly, because it needs to include pointer to `callback` created using `wasm::Func::make()` which includes pointer to said `FuncData`. I could set `raw_func` in constructor, but I left it here for readability.

I'm fine changing things if you can provide code snippets, but I'm not sure what changes do you want to see, exactly.

Reply:
Ah, I didn't see the recursive reference in the lambda.  LGTM